### PR TITLE
Fixed detection of identifier getter with PHP 7 return type

### PR DESCRIPTION
**Pull request by janlanger**

### CODE REVIEW

File: lib/Doctrine/Common/Proxy/ProxyGenerator.php [hunk 1]
```diff
@@ -37,7 +37,7 @@ class ProxyGenerator
      * Used to match very simple id methods that don't need
      * to be decorated since the identifier is known.
      */
-    const PATTERN_MATCH_ID_METHOD = '((public\s+)?(function\s+%s\s*\(\)\s*)\s*{\s*return\s*\$this->%s;\s*})i';
+    const PATTERN_MATCH_ID_METHOD = '((public\s+)?(function\s+%s\s*\(\)\s*)\s*(?::\s*[\w\\\\]+\s*)?{\s*return\s*\$this->%s;\s*})i';
```
Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: php type identifiers follow the regex `'/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]+$/'`

Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: Also: needs a test case

Comment (janlanger) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: @Ocramius But there can be more identifiers separated by backslash - namespaces. This version is simple, I could extend it to include `\x7f-\xff` or use strict version like `(?::\s*\\?[a-z_\x7f-\xff][\w\x7f-\xff]*(?:\\[a-z_\x7f-\xff][\w\x7f-\xff]*)*\s*)`. But is strict really necessary? The checked code have to be valid PHP anyway and there cannot be any code between closing `)` and opening `{`

I'll try to prepare test, is there any existing testcase for this feature? I didn't find any...

Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: @janlanger well, general purpose lib => general purpose solution :-)

In think tests for this particular scenario are somewhere in https://github.com/doctrine/common/blob/faf7d81273eeffb84d524006daf85d37a14ac42a/tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php

**Comment by janlanger** [comment]
@Ocramius updated

File: lib/Doctrine/Common/Proxy/ProxyGenerator.php [hunk 2]
```diff
@@ -37,7 +37,7 @@ class ProxyGenerator
      * Used to match very simple id methods that don't need
      * to be decorated since the identifier is known.
      */
-    const PATTERN_MATCH_ID_METHOD = '((public\s+)?(function\s+%s\s*\(\)\s*)\s*{\s*return\s*\$this->%s;\s*})i';
+    const PATTERN_MATCH_ID_METHOD = '((public\s+)?(function\s+%s\s*\(\)\s*)\s*(?::\s*\\\\?[a-z_\x7f-\xff][\w\x7f-\xff]*(?:\\\\[a-z_\x7f-\xff][\w\x7f-\xff]*)*\s*)?{\s*return\s*\$this->%s;\s*})i';
```
Comment (Ener-Getick) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: what about using ``[^\{]*`` instead ?

File: tests/Doctrine/Tests/Common/Proxy/ProxyLogicTypehintsTest.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+use Doctrine\Common\Proxy\ProxyGenerator;
+use Doctrine\Common\Proxy\Proxy;
+use Doctrine\Common\Proxy\Exception\UnexpectedValueException;
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use PHPUnit_Framework_TestCase;
+
+/**
+ * Test that identifier getter does not cause lazy loading. These tests make assumptions about the structure of LazyLoadableObjectWithTypehints
+ *
+ * @author Marco Pivetta <ocramius@gmail.com>
```
Comment (Ener-Getick) [inline tests/Doctrine/Tests/Common/Proxy/ProxyLogicTypehintsTest.php]: should be your own name ;)

File: lib/Doctrine/Common/Proxy/ProxyGenerator.php (hunk 2 above)
Comment (janlanger) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: @Ener-Getick That would not be "general purpose solution" :) See https://github.com/doctrine/common/pull/720#discussion_r65870837 and following,

Comment (Ener-Getick) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: I would prefer simplicity over strictness in this case but looks good anyway ☺

Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: Correctness is required for generic-purpose libraries, or else there will be dragons.

File: tests/Doctrine/Tests/Common/Proxy/ProxyLogicTypehintsTest.php (hunk 3 above)
Comment (janlanger) [inline tests/Doctrine/Tests/Common/Proxy/ProxyLogicTypehintsTest.php]: The code is basically modified copy of ProxyLogicTest. I added myself as second author :)

**Comment by janlanger** [comment]
Updated & rebased on master

**Comment by Ocramius** [comment]
@janlanger merging, thanks!

**Comment by solcik** [comment]
Hello,
as this pull request accepted option to add a return type for identifier getter method.
Is it possible to accept typecasting in regex **PATTERN_MATCH_ID_METHOD**?

Getter like this one, will generate proxy below.

```php
public function getId(): int
{
    return $this->id;
}
```

```php
/**
 * {@inheritDoc}
 */
public function getId(): int
{
    if ($this->__isInitialized__ === false) {
        return (int)  parent::getId();
    }


    $this->__initializer__ && $this->__initializer__->__invoke($this, 'getId', []);

    return parent::getId();
}
```

So when I call **::getId()** in not initialized entity, it throws **TypeError**, because **$this->id** is *string*.

```
TypeError
Return value of ...::getId() must be of the type integer, string returned
```

So It would be really nice, if I could typecast identifier and not end up with

```php
public function getId(): int
{
    return (int) $this->id;
}
```

```php
/**
 * {@inheritDoc}
 */
public function getId(): int
{
    $this->__initializer__ && $this->__initializer__->__invoke($this, 'getId', []);

    return parent::getId();
}
```

**Comment by Ocramius** [comment]
David, that sounds super-broken. The ORM already casts identifiers to their
mapped types: casting yourself too indicates a design problem in your
codebase first, and the ORM shouldn't help you in supporting bad practices.

**Pull request merged**